aptos-move/aptos-vm/src/sharded_block_executor/sharded_aggregator_service.rs

```diff
@@ -0,0 +1,251 @@
+
+        let delta2 = DeltaU128 {
+            delta: 50,
+            is_positive: false,
+        };
+        assert_eq!(delta2.add_delta(50), 0);
+    }
+
+    {
+        // test all cases for ops::Add
+        let delta1 = DeltaU128 {
+            delta: 100,
+            is_positive: true,
+        };
+        let delta2 = DeltaU128 {
+            delta: 50,
+            is_positive: false,
+        };
+        let delta3 = DeltaU128 {
+            delta: 100,
+            is_positive: true,
+        };
+        // checks for [pos > neg]; [pos, pos]
+        let mut delta_sum = delta1 + delta2 + delta3;
+        assert_eq!(delta_sum.delta, 150);
+        assert!(delta_sum.is_positive);
+
+        let delta4 = DeltaU128 {
+            delta: 500,
+            is_positive: false,
+        };
+        let delta5 = DeltaU128 {
+            delta: 200,
+            is_positive: false,
+        };
+        // checks for [neg > pos]; [neg, neg]
+        delta_sum = delta_sum + delta4 + delta5;
+        assert_eq!(delta_sum.delta, 550);
+        assert!(!delta_sum.is_positive);
+    }
+}
+
+pub fn aggregate_and_update_total_supply<S: StateView>(
+    sharded_output: &mut Vec<Vec<Vec<TransactionOutput>>>,
+    global_output: &mut [TransactionOutput],
+    state_view: &S,
+    executor_thread_pool: Arc<rayon::ThreadPool>,
+) {
+    let num_shards = sharded_output.len();
+    let num_rounds = sharded_output[0].len();
+
+    // The first element is 0, which is the delta for shard 0 in round 0. +1 element will contain
+    // the delta for the global shard
+    let mut aggr_total_supply_delta = vec![DeltaU128::default(); num_shards * num_rounds + 1];
+
+    // No need to parallelize this as the runtime is O(num_shards)
+    // TODO: Get this from the individual shards while getting 'sharded_output'
+    sharded_output
+        .iter()
+        .enumerate()
+        .for_each(|(shard_id, shard_output)| {
+            for (round, txn_outputs) in shard_output.iter().enumerate() {
+                // Though we expect all the txn_outputs to have total_supply, there can be
+                // exceptions like 'block meta' (first txn in the block) and 'chkpt info' (last txn
+                // in the block) which may not have total supply. Hence we iterate till we find the
+                // last txn with total supply.
+                for txn in txn_outputs.iter().rev() {
+                    if let Some(last_txn_total_supply) = txn.write_set().get_total_supply() {
+                        aggr_total_supply_delta[round * num_shards + shard_id + 1] =
+                            DeltaU128::get_delta(last_txn_total_supply, TOTAL_SUPPLY_AGGR_BASE_VAL);
+                        break;
+                    }
+                }
+            }
+        });
+
+    for idx in 1..aggr_total_supply_delta.len() - 1 {
+        aggr_total_supply_delta[idx + 1] =
+            aggr_total_supply_delta[idx + 1] + aggr_total_supply_delta[idx];
```

Comment:
Using += has some rust lang issues.
`209 |         aggr_total_supply_delta[idx + 1] += (aggr_total_supply_delta[idx]);
    |         -------------------------------------^^^^^^^^^^^^^^^^^^^^^^^------
    |         |                                    |
    |         |                                    immutable borrow occurs here
    |         mutable borrow occurs here
    |         mutable borrow later used here
    |`
    
    https://www.reddit.com/r/rust/comments/offqy6/cannot_borrow_self_as_mutable_because_it_is_also/

I will have to use another local variable to get around this, and that will require 2 lines of code instead of 1

Reply:
:trollface:  I'd prefer two lines but doesn't matter much.

Reply:
As per the change from addressing another comment, we now don't need '+=' operator